hazelcast/src/main/java/com/hazelcast/sql/impl/exec/UpstreamState.java

```diff
@@ -0,0 +1,147 @@
+import com.hazelcast.sql.impl.row.EmptyRowBatch;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+
+import java.util.Iterator;
+
+import static com.hazelcast.sql.impl.exec.IterationResult.FETCHED_DONE;
+import static com.hazelcast.sql.impl.exec.IterationResult.WAIT;
+
+/**
+ * Upstream state.
+ */
+public class UpstreamState implements Iterable<Row> {
+    /** Upstream operator. */
+    private final Exec upstream;
+
+    /** Iterator over the current batch. */
+    private final UpstreamIterator iter;
+
+    /** Current batch returned from the upstream. */
+    private RowBatch currentBatch = EmptyRowBatch.INSTANCE;
+
+    /** Current position. */
+    private int currentBatchPos;
+
+    /** Last returned state. */
+    private IterationResult state;
+
+    public UpstreamState(Exec upstream) {
+        this.upstream = upstream;
+
+        iter = new UpstreamIterator();
+    }
+
+    /**
+     * Try advancing the upstream.
+     *
+     * @return {@code True} if the caller may try iteration over results; {@code false} if the caller should give
+     * up execution and wait.
+     */
+    public boolean advance() {
+        // If some data is available still, do not do anything, just return the previous result.
+        if (currentBatchPos < currentBatch.getRowCount()) {
+            return true;
+        }
+
+        // If the upstream is exhausted, just return "done" flag.
+        if (state == FETCHED_DONE) {
+            return true;
+        }
+
+        // Otherwise poll the upstream.
+        state = upstream.advance();
+
+        switch (state) {
+            case FETCHED_DONE:
+            case FETCHED:
+                currentBatch = upstream.currentBatch();
+                assert currentBatch != null;
+
+                currentBatchPos = 0;
+
+                return true;
+
+            default:
+                assert state == WAIT;
+
+                currentBatch = EmptyRowBatch.INSTANCE;
+                currentBatchPos = 0;
+
+                return false;
+        }
+    }
+
+    public void setup(QueryFragmentContext ctx) {
+        upstream.setup(ctx);
+    }
+
+    public RowBatch consumeBatch() {
+        if (currentBatchPos != 0) {
+            throw HazelcastSqlException.error("Batch can be consumed only as a whole: " + upstream);
+        }
+
+        RowBatch batch = currentBatch;
+
+        currentBatchPos = batch.getRowCount();
+
+        return batch;
+    }
+
+    /**
+     * @return {@code True} if no more results will appear in future.
+     */
+    public boolean isDone() {
+        return state == FETCHED_DONE && !iter.hasNext();
+    }
+
+    @SuppressWarnings("NullableProblems")
+    @Override
+    public Iterator<Row> iterator() {
+        return iter;
+    }
+
+    /**
+     * Return next row in the current batch if it exists.
+     *
+     * @return Next row or {@code null}.
+     */
+    public Row nextIfExists() {
```

Comment:
Why not to use standard `hasNext()/next() `methods semantics from `Iterator` instead of introducing `isDone/nextIfExists`?

Reply:
This is just a convenient shortcut for iterator hasNext/next. You may see usages of it in the prototype branch.

Reply:
OK, but why would you introduce new names?

Reply:
Because it has different semantics - it doesn't throw an exception is there is no row.

Reply:
I guess, you have `isDone()` to avoid calling this method when there is no new Row to consume. Anyway, it is a matter of taste.